File: Tutorials/ObjC/PerpetualLicense/Download2GoHelloWorld/DemoPlayerViewController.m
//
//  DemoPlayerViewController.m
//  Download2GoHelloWorld-Objc
//

#import "DemoPlayerViewController.h"

@implementation DemoPlayerViewController

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self play];
    
    // Redirect errors from FairPlay licensing here so we can handle reporting errors to the user.
    AVURLAsset* asset = (AVURLAsset*)self.playerView.player.currentItem.asset;
    if( [[asset.resourceLoader.delegate class]conformsToProtocol:@protocol(VirtuosoAVAssetResourceLoaderDelegate)] )
    {
        id<VirtuosoAVAssetResourceLoaderDelegate> loader = (id<VirtuosoAVAssetResourceLoaderDelegate>)asset.resourceLoader.delegate;
        loader.errorHandler = self;  // see method resourceLoaderDelegate below
    }
}

- (void)resourceLoaderDelegate:(id<VirtuosoAVAssetResourceLoaderDelegate>)delegate generatedError:(NSError *)error
{
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"License Error" message:@"An error was encountered playing this content.  Please try again later." preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [alert addAction:okAction];
        [self presentViewController:alert animated:YES completion:nil];
    });
}

@end

File: Tutorials/Swift/Example17/Example17/DemoPlayerViewController.swift
//
//  DemoPlayerViewController.swift
//  Example17
//
//  Created by Penthera on 10/13/21.
//

import UIKit

class DemoPlayerViewController: VirtuosoPlayerViewController, VirtuosoAVAssetResourceLoaderDelegateErrorHandler {
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
        self.play()
        
        // Redirect errors from FairPlay licensing here so we can handle reporting errors to the user.
        let asset = self.playerView.player?.currentItem?.asset as! AVURLAsset       
        if let loader = asset.resourceLoader.delegate as? VirtuosoAVAssetResourceLoaderDelegate {
            loader.errorHandler = self
        }
    }

    func resourceLoaderDelegate(_ delegate: VirtuosoAVAssetResourceLoaderDelegate, generatedError error: Error?)
    {
        DispatchQueue.main.async {
            let alert = UIAlertController(title: "License Error", message: "An error was encountered playing this content.  Please try again later.", preferredStyle: .alert)
            let okAction = UIAlertAction(title: "OK", style: .default, handler: nil)
            alert.addAction(okAction)
            self.present(alert, animated: true, completion: nil)
        }
    }
}

